File: TeenAstroASCOM/TeenAstroASCOM/SetupDialogForm.vb
Imports System.Windows.Forms
Imports System.Runtime.InteropServices
Imports ASCOM.Utilities
Imports ASCOM.TeenAstro

<ComVisible(False)>
Public Class SetupDialogForm

  Private Sub OK_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OK_Button.Click ' OK button event handler
    ' Persist new values of user settings to the ASCOM profile
    If RadioButtonSP.Checked Then
      Telescope.mcomPort = ComboBoxComPort.SelectedItem
      Telescope.mInterface = "COM"
    ElseIf RadioButtonIP.Checked Then
      Telescope.mIP = TextBoxIP.Text
      Telescope.mPort = TextBoxPort.Text
      Telescope.mInterface = "IP"

      If Not System.Net.IPAddress.TryParse(Telescope.mIP, Telescope.mobjectIP) Then
        MsgBox(Telescope.mIP + " is not a valid IP Address")
        Return
      End If
    End If
    Telescope.mtraceState = chkTrace.Checked
    Me.DialogResult = System.Windows.Forms.DialogResult.OK
    Me.Close()
  End Sub

  Private Sub Cancel_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Cancel_Button.Click 'Cancel button event handler
    Me.DialogResult = System.Windows.Forms.DialogResult.Cancel
    Me.Close()
  End Sub

  Private Sub ShowAscomWebPage(ByVal sender As System.Object, ByVal e As System.EventArgs)
    ' Click on ASCOM logo event handler
    Try
      System.Diagnostics.Process.Start("http://ascom-standards.org/")
    Catch noBrowser As System.ComponentModel.Win32Exception
      If noBrowser.ErrorCode = -2147467259 Then
        MessageBox.Show(noBrowser.Message)
      End If
    Catch other As System.Exception
      MessageBox.Show(other.Message)
    End Try
  End Sub

  Private Sub SetupDialogForm_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load ' Form load event handler
    ' Retrieve current values of user settings from the ASCOM Profile
    InitUI()
  End Sub

  Private Sub InitUI()

    Dim assembly As System.Reflection.Assembly = System.Reflection.Assembly.GetExecutingAssembly()
    Dim fvi As FileVersionInfo = FileVersionInfo.GetVersionInfo(assembly.Location)
    Dim version As String = fvi.FileVersion
    Me.Text = "TeenAstro Telescope " + version
    chkTrace.Checked = Telescope.mtraceState
    TextBoxIP.Text = Telescope.mIP
    TextBoxPort.Text = Telescope.mPort
    RadioButtonSP.Checked = Telescope.mInterface = "COM"
    RadioButtonIP.Checked = Telescope.mInterface = "IP"
    ' set the list of com ports to those that are currently available
    ComboBoxComPort.Items.Clear()
    ComboBoxComPort.Items.AddRange(System.IO.Ports.SerialPort.GetPortNames())       ' use System.IO because it's static
    ' select the current port if possible
    If ComboBoxComPort.Items.Contains(Telescope.mcomPort) Then
      ComboBoxComPort.SelectedItem = Telescope.mcomPort
    End If
  End Sub


  Private Sub RadioButtonIP_CheckedChanged(sender As Object, e As EventArgs) Handles RadioButtonIP.CheckedChanged
    TextBoxIP.Enabled = RadioButtonIP.Checked
    TextBoxPort.Enabled = RadioButtonIP.Checked
  End Sub

  Private Sub RadioButtonSP_CheckedChanged(sender As Object, e As EventArgs) Handles RadioButtonSP.CheckedChanged
    ComboBoxComPort.Enabled = RadioButtonSP.Checked
  End Sub

  Private Sub TextBoxPort_KeyPress(sender As Object, e As KeyPressEventArgs) Handles TextBoxPort.KeyPress
    If Not Char.IsDigit(e.KeyChar) And Not Char.IsControl(e.KeyChar) Then
      e.Handled = True
    End If
  End Sub

End Class

File: TeenAstroASCOM_V7/TeenAstroASCOM_V7/TelescopeDriver/SetupDialogForm.cs
using ASCOM.Utilities;
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Windows.Forms;

namespace ASCOM.TeenAstro.Telescope
{
  [ComVisible(false)] // Form not registered for COM!
  public partial class SetupDialogForm : Form
  {
    const string NO_PORTS_MESSAGE = "No COM ports found";
    TraceLogger tl; // Holder for a reference to the driver's trace logger

    public SetupDialogForm(TraceLogger tlDriver)
    {
      InitializeComponent();

      // Save the provided trace logger for use within the setup dialogue
      tl = tlDriver;

      // Initialise current values of user settings from the ASCOM Profile
      InitUI();
    }


    private void CmdCancel_Click(object sender, EventArgs e) // Cancel button event handler
    {
      Close();
    }

    private void BrowseToAscom(object sender, EventArgs e) // Click on ASCOM logo event handler
    {
      try
      {
        System.Diagnostics.Process.Start("https://ascom-standards.org/");
      }
      catch (Win32Exception noBrowser)
      {
        if (noBrowser.ErrorCode == -2147467259)
          MessageBox.Show(noBrowser.Message);
      }
      catch (Exception other)
      {
        MessageBox.Show(other.Message);
      }
    }

    private void InitUI()
    {

      System.Reflection.Assembly assembly = System.Reflection.Assembly.GetExecutingAssembly();
      FileVersionInfo fvi = FileVersionInfo.GetVersionInfo(assembly.Location);
      String version = fvi.FileVersion;
      this.Text = "TeenAstro Telescope " + version;

      // Set the trace checkbox
      chkTrace.Checked = tl.Enabled;
      TextBoxIP.Text = TelescopeHardware.IP;
      TextBoxPort.Text = TelescopeHardware.Port.ToString();
      RadioButtonSP.Checked = TelescopeHardware.Interface == "COM";
      RadioButtonIP.Checked = TelescopeHardware.Interface == "IP";
      // set the list of COM ports to those that are currently available
      comboBoxComPort.Items.Clear(); // Clear any existing entries
      using (Serial serial = new Serial()) // User the Se5rial component to get an extended list of COM ports
      {
        comboBoxComPort.Items.AddRange(serial.AvailableCOMPorts);
      }

      // If no ports are found include a message to this effect
      if (comboBoxComPort.Items.Count == 0)
      {
        comboBoxComPort.Items.Add(NO_PORTS_MESSAGE);
        comboBoxComPort.SelectedItem = NO_PORTS_MESSAGE;
      }

      // select the current port if possible
      if (comboBoxComPort.Items.Contains(TelescopeHardware.comPort))
      {
        comboBoxComPort.SelectedItem = TelescopeHardware.comPort;
      }

      tl.LogMessage("InitUI", $"Set UI controls to Trace: {chkTrace.Checked}, COM Port: {comboBoxComPort.SelectedItem}");
    }

    private void SetupDialogForm_Load(object sender, EventArgs e)
    {
      // Bring the setup dialogue to the front of the screen
      if (WindowState == FormWindowState.Minimized)
        WindowState = FormWindowState.Normal;
      else
      {
        TopMost = true;
        Focus();
        BringToFront();
        TopMost = false;
      }
    }

    private void OK_Button_Click(object sender, EventArgs e)
    {
      // Place any validation constraint checks here and update the state variables with results from the dialogue

      tl.Enabled = chkTrace.Checked;
      if (RadioButtonSP.Checked)
      {
        // Update the COM port variable if one has been selected
        if (comboBoxComPort.SelectedItem is null) // No COM port selected
        {
          tl.LogMessage("Setup OK", $"New configuration values - COM Port: Not selected");
        }
        else if (comboBoxComPort.SelectedItem.ToString() == NO_PORTS_MESSAGE)
        {
          tl.LogMessage("Setup OK", $"New configuration values - NO COM ports detected on this PC.");
        }
        else // A valid COM port has been selected
        {
          tl.LogMessage("Setup OK", $"New configuration values - COM Port: {comboBoxComPort.SelectedItem}");
          TelescopeHardware.comPort = comboBoxComPort.SelectedItem.ToString();
          TelescopeHardware.Interface = "COM";
        }
      }
      else if (RadioButtonIP.Checked)
      {
        TelescopeHardware.IP = TextBoxIP.Text;
        TelescopeHardware.Port = Convert.ToInt16(TextBoxPort.Text);
        TelescopeHardware.Interface = "IP";
      }
      this.DialogResult = DialogResult.OK;
      this.Close();
    }

    private void PictureBoxTeenAstro_Click(object sender, EventArgs e)
    {

    }

    private void RadioButtonIP_CheckedChanged(object sender, EventArgs e)
    {
      TextBoxIP.Enabled = RadioButtonIP.Checked;
      TextBoxPort.Enabled = RadioButtonIP.Checked;
    }
    private void RadioButtonSP_CheckedChanged(object sender, EventArgs e)
    {
      comboBoxComPort.Enabled = RadioButtonSP.Checked;
    }
    private void TextBoxPort_KeyPress(object sender, KeyPressEventArgs e)
    {
      if (!char.IsDigit(e.KeyChar) & !char.IsControl(e.KeyChar))
      {
        e.Handled = true;
      }
    }
    private void chkTrace_CheckedChanged(object sender, EventArgs e)
    {

    }

    private void Cancel_Button_Click(object sender, EventArgs e)
    {
      this.Close();
    }
  }
}